### ISSUE TITLE
Provide error codes

### ISSUE THREAD

**Issue by codinfox** [rethinkdb/rethinkdb#4223]
In case of error, please provide an error code besides the error message.

```js
r.dbCreate('test').run(conn).then(...).error(function(err) {
    console.log(err);
});
```

The absence of error codes makes it harder to handle errors. In the case above, how can I know if the failure of `r.dbCreate()` is due to internal error or because of an existent database with the same name? I don't think string comparison of `err.msg` is a good idea. 

Therefore, please provide error codes.

**Comment by danielmewes** [comment]
Thanks for opening an issue about this @codinfox .
We have been planning to do this, but as far as I know haven't had an issue for it yet.

Error codes are one option. Another one is having a proper error type hierarchy with specific error types.

**Comment by danielmewes** [comment]
Renamed from "Provide error code for js driver" to "Provide error codes" since the issue is not specific to the JS driver, but requires both server and driver changes in all languages.

**Comment by codinfox** [comment]
A proper error type hierarchy is even better! Thank you @danielmewes . Looking forward to it!

**Comment by danielmewes** [comment]
This is partially implemented in 2.1 (see https://github.com/rethinkdb/rethinkdb/issues/4544). Keeping https://github.com/rethinkdb/rethinkdb/issues/559 for more fine-grained error types in the future.
